trust-quorum/tests/coordinator.rs

```diff
@@ -0,0 +1,646 @@
+    fn assert_envelopes_after_coordinate_reconfiguration(
+        &self,
+        outbox: &[Envelope],
+    ) -> Result<(), TestCaseError> {
+        let sut = &self.sut;
+        let msg = &self
+            .model
+            .coordinator_state
+            .as_ref()
+            .ok_or(TestCaseError::fail(
+                "Model should have a coordinator state",
+            ))?
+            .msg;
+
+        let config = sut.persistent_state.configuration(msg.epoch).unwrap();
+
+        // Ensure the members of the configuration match the model msg
+        prop_assert_eq!(
+            &msg.members,
+            &config.members.keys().cloned().collect()
+        );
+
+        // The coordinator should send messages to every node but itself
+        assert_eq!(outbox.len(), config.members.len() - 1);
+        for envelope in outbox {
+            assert_matches!(
+                &envelope.msg,
+                PeerMsg::Prepare(PrepareMsg { config: prepare_config, .. }) => {
+                    assert_eq!(*config, *prepare_config);
+                }
+            );
+            prop_assert_eq!(&envelope.from, &config.coordinator);
+
+            // We don't send a prepare to ourselves. We put it in our
+            // `PersistentState` directly.
+            prop_assert_ne!(&envelope.to, &config.coordinator);
+        }
+
+        Ok(())
+    }
+
+    fn send(&mut self, envelopes: impl Iterator<Item = Envelope>) {
+        for envelope in envelopes {
+            let msgs = self.network_msgs.entry(envelope.to).or_default();
+            msgs.push(envelope.msg);
+        }
+    }
+}
+
+// A high-level set of actions to drive the state of a peer forward.
+#[derive(Debug, Arbitrary)]
+pub enum Action {
+    /// For each indexed member deliver an in-flight network msg if there is
+    /// one.
+    ///
+    /// The indexes here are used to index into the `PlatformIds` of
+    /// `test_state.network_msgs`.
+    ///
+    /// We may deliver more than one message to each member.
+    #[weight(50)]
+    DeliverMsgs(Vec<Index>),
+
+    /// For each indexed member drop an in-flight network msg if there is
+    /// one.
+    ///
+    /// The indexes here are used to index into the `PlatformIds` of
+    /// `test_state.network_msgs`.
+    ///
+    /// We may drop more than one message to each member, which is why we
+    /// use a Vec. We don't worry about shuffling messsages so they are out of
+    /// order, because we use stream oriented transport.
+    //#[weight(3)]
+    //DropMessages(Vec<Index>),
+
+    /// Fake a reply for a delivered message.
+    ///
+    /// The indexes here are used to index into the `PlatformIds` of
+    /// `test_state.delivered_msgs`.
+    #[weight(45)]
+    Reply(Vec<Index>),
+    // Start a reconfiguration that has a possibility of succeeding
+    // given delivered messages.
+    //
+    // It will never fail a validation check.
+    //#[weight(2)]
+    //StartReconfiguration(GeneratedConfiguration),
```

Comment:
dead code -- remove?

Reply:
Normally, absolutely, but I'm using it in the next PR, coming soon.